### PR TITLE
chakra 1.4.0 (new formula)

### PR DESCRIPTION
**Pull request by mathiasbynens**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Ref. https://github.com/Microsoft/ChakraCore/issues/2278

### CODE REVIEW

**Comment by mathiasbynens** [comment]
@obastemur [mentioned](https://github.com/Microsoft/ChakraCore/issues/2417#issuecomment-275652850) the issue could be this:

> IIRC Homebrew pushes for recent libc++ for osx 10.9+..
> 
> While installing with homebrew, you should use `--env=std` argument with brew install ... command. The rest should be just fine.

In that case, how can I force Homebrew to default to old libc++ for this formula, i.e. act as if `--env=std` was always supplied?

File: Formula/chakra.rb
Comment (MikeMcQuaid) [inline Formula/chakra.rb]: Why `BuildLinux`?

Comment (ilovezfs) [inline Formula/chakra.rb]: It's hard coded: https://ci.dot.net/job/Microsoft_ChakraCore/job/master/job/osx_osx_release_static/lastBuild/console

**Comment by ilovezfs** [comment]
The problem is not the c++ library, it's the optimization. Adding `ENV.O3` to the install method should fix it for you.

File: Formula/chakra.rb
Comment (MikeMcQuaid) [inline Formula/chakra.rb]: Gentle nudge that upstream might want to consider renaming at some point 😉

File: Formula/chakra.rb
Comment (MikeMcQuaid) [inline Formula/chakra.rb]: I'd populate an `args` array here instead as that's more typical formula form and inline `icu4c` as you're only using that variable once

**Comment by ilovezfs** [comment]
This has a case insensitive file system bug, which can be fixed with
```
inreplace "pal/src/locale/unicode.cpp", "corefoundation", "CoreFoundation"
```

File: Formula/chakra.rb
Comment (ilovezfs) [inline Formula/chakra.rb]: should be a release build, not a test build

File: Formula/chakra.rb
Comment (ilovezfs) [inline Formula/chakra.rb]: This should be `bin.install` not `prefix.install`

**Comment by ilovezfs** [comment]
I suggest the following:

```
class Chakra < Formula
  desc "The core part of the Chakra Javascript engine that powers Microsoft Edge"
  homepage "https://github.com/Microsoft/ChakraCore"
  url "https://github.com/Microsoft/ChakraCore/archive/v1.4.0.tar.gz"
  sha256 "f6b4a3bce9e69bb3ea619a82805460906ef53d4f9da98b51e554dd48d6cdfdb3"

  depends_on "cmake" => :build
  depends_on "icu4c" => :optional

  def install
    ENV.O3

    # Fix build failure on case-sensitive file systems
    inreplace "pal/src/locale/unicode.cpp", "corefoundation", "CoreFoundation"

    args = ["--static"]
    if build.with? "icu4c"
      args << "--icu=#{Formula["icu4c"].opt_include}"
    else
      args << "--no-icu"
    end
    system "./build.sh", *args

    bin.install "BuildLinux/Release/ch" => "chakra"
  end

  test do
    (testpath/"test.js").write("print('Hello world!');\n")
    assert_equal "Hello world!", shell_output("#{bin}/chakra test.js").chomp
  end
end
```

File: Formula/chakra.rb
Comment (mathiasbynens) [inline Formula/chakra.rb]: What would inlining look like exactly? Would `"--icu=#{Formula['icu4c'].opt_include}"` work? (I’m not a Ruby programmer 😊)

File: Formula/chakra.rb
Comment (mathiasbynens) [inline Formula/chakra.rb]: The test build enables command-line flags so most of the time people will want a test build. I’ll make this an option then

File: Formula/chakra.rb
Comment (mathiasbynens) [inline Formula/chakra.rb]: Ah never mind, @ilovezfs already shows this below:

```
"--icu=#{Formula["icu4c"].opt_include}"
```

File: Formula/chakra.rb
Comment (obastemur) [inline Formula/chakra.rb]: ``--target`` argument is added to set custom path and will be part of release 2.0. We also don't like `BuildLinux` naming. I have to admit, we've been lazy on changing it..

BTW; `BuildLinux/Release/ch` is a wrong path for `test-build`. It should be `BuildLinux/Test/ch`

**Comment by mathiasbynens** [comment]
Thanks for the helpful reviews, @ilovezfs and @MikeMcQuaid! ♥ I’ve updated the PR.

**Comment by ilovezfs** [comment]
@MikeMcQuaid are we permitting the use of libstdc++ in new formulae these days?

**Comment by MikeMcQuaid** [comment]
@ilovezfs not intentionally but in this case we probably wouldn't have noticed and there's no obvious hacks happening. 
@mathiasbynens Why is `libc++` not supported, any ideas?

**Comment by mathiasbynens** [comment]
> @mathiasbynens Why is `libc++` not supported, any ideas?

I have no idea — maybe @obastemur does? (Note: I am not affiliated with Microsoft or the ChakraCore team.)

**Comment by obastemur** [comment]
> @mathiasbynens Why is libc++ not supported, any ideas?

https://github.com/Microsoft/ChakraCore/issues/2417#issuecomment-275849276

**Comment by MikeMcQuaid** [comment]
Will wait until @ilovezfs's conversation there has concluded and leave it up to him whether we merge this given the stdlib situation.

**Comment by ilovezfs** [comment]
-Os issue is fixed. You can drop the ENV.O3, by adding this patch from upstream:

```
  patch do
    url "https://github.com/obastemur/ChakraCore/commit/54874cfbe139daad5ae821c5e2710b6764632bd4.patch"
    sha256 "8f08b9708e43b9f3324187573455f3a1c01fe6861e39f7f909a60aced314c0ae"
  end
```

However, the deployment target/mmacosx-version-min and C++ issues remain unresolved: https://github.com/Microsoft/ChakraCore/issues/2417#issuecomment-276072593

**Comment by mathiasbynens** [comment]
Patch updated!

**Comment by mathiasbynens** [comment]
The build now fails on Yosemite:

```
[100%] Linking CXX executable ch
Undefined symbols for architecture x86_64:
  "Js::JavascriptString::ConcatDestructive(Js::JavascriptString*)", referenced from:
      Js::JavascriptMath::AddLeftDead(void*, void*, Js::ScriptContext*, Js::JavascriptNumber*) in libChakra.Jsrt.a(JavascriptMathOperators.cpp.o)
      Js::JavascriptMath::Add_FullHelper(void*, void*, Js::ScriptContext*, Js::JavascriptNumber*, bool) in libChakra.Jsrt.a(JavascriptMathOperators.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/ch/ch] Error 1
make[1]: *** [bin/ch/CMakeFiles/ch.dir/all] Error 2
make: *** [all] Error 2
See error details above. Exit code was 2
```

cc @obastemur

**Comment by obastemur** [comment]
@mathiasbynens it works on master.

we will have our first official cross-platform release soon, we may target that instead of 1.4.0? In the meantime, I will make sure we have this scenario covered in our CI. So, we wouldn't break it going forward.

**Comment by mathiasbynens** [comment]
Can’t wait for that release!

File: Formula/chakra.rb
Comment (MikeMcQuaid) [inline Formula/chakra.rb]: Can we backport this commit as a patch so it gets removed on the next release?

File: Formula/chakra.rb
Comment (MikeMcQuaid) [inline Formula/chakra.rb]: Similarly here.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

File: Formula/chakra.rb
Comment (ilovezfs) [inline Formula/chakra.rb]: Nope. It causes the build to fail on Yosemite.

File: Formula/chakra.rb
Comment (ilovezfs) [inline Formula/chakra.rb]: Yes, this one could be applied as a patch.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by ilovezfs** [comment]
@mathiasbynens Thanks for the new formula! :new:

**Comment by mathiasbynens** [comment]
Thanks for the help! 👍

**Pull request closed without merging**
